### ISSUE TITLE
BL/WL handling does not match R package behavior

### ISSUE THREAD

**Issue by rbartelme** [erdogant/bnlearn#23]
I've been trying to understand the `list` argument for the BL/WL handling in this python port vs. the `bnlearn` R package. The R package takes information in what R deems a matrix, but python may call it a character array. 

Example of a whitelist in R's `bnlearn`:

```Rscript
wl <- matrix(c("cultivar", "canopy_height", "gdd", "canopy_height",
              "precip_cumulative", "canopy_height"),
             ncol = 2, byrow = TRUE,
             dimnames = list(NULL, c("from", "to")))
```

It's unclear from the example in the documentation if this package treats this similarly with the list argument:

```python
import bnlearn
# Load example mixed dataset
df_raw = bnlearn.import_example(data='titanic')

# Convert to onehot
dfhot, dfnum = bnlearn.df2onehot(df_raw)

# Structure learning by including only 'Survived','Pclass','Sex','Embarked','Parch'.
DAG_filtered = bnlearn.structure_learning.fit(dfnum, methodtype='hc', bw_list_method='filter', white_list=['Survived','Pclass','Sex','Embarked','Parch'])

# Structure learning by enforcing variables 'Survived','Pclass','Sex','Embarked','Parch'.
DAG_enforced = bnlearn.structure_learning.fit(dfnum, methodtype='hc', bw_list_method='enforce', white_list=['Survived','Pclass','Sex','Embarked','Parch'])

# Plot
Gf = bnlearn.plot(DAG_filtered)
Ge = bnlearn.plot(DAG_enforced)

```

Or would I need to structure this by hard coding a DAG, then learning the additional nodes from a dataset?

Like this:

```python
# Import the library
import bnlearn

# Define the network structure
edges = [('Cloudy', 'Sprinkler'),
         ('Cloudy', 'Rain'),
         ('Sprinkler', 'Wet_Grass'),
         ('Rain', 'Wet_Grass')]

# Make the actual Bayesian DAG
DAG = bnlearn.make_DAG(edges)

```

**Comment by erdogant** [comment]
The BL/WL parameter (with filter) in Python takes a list and in case of blacklist it excludes the variables with that name. In case of whitelist it keeps the specified onces. Im not sure anymore how the R package handles this.

**Comment by linghelenchen** [comment]
I encounter the same problem. After spending a whole day until I saw this question.
In bnlearn R package, blacklist and whitelist are for edges, not variables. However, in bnlearn Python package, they are for variables. 
I actually need the whitelist and blacklist of edges. Seems like I have to change to R. So sad!!!

**Comment by linghelenchen** [comment]
> The BL/WL parameter (with filter) in Python takes a list and in case of blacklist it excludes the variables with that name. In case of whitelist it keeps the specified onces. Im not sure anymore how the R package handles this.
Thanks for implementing this package! It helps me a lot. However, I also have the same question.
Does bnlearn can blacklist and whitelist edges like R packages. I think they may more useful than to blacklist and whitelist variables.

**Comment by erdogant** [comment]
Thank you for the suggestion. Can you maybe demonstrate it with a small example how the R package handles the BW listings? I could for example include another parameter to filter for nodes or edges.

**Comment by linghelenchen** [comment]
> Thank you for the suggestion. Can you maybe demonstrate it with a small example how the R package handles the BW listings? I could for example include another parameter to filter for either nodes or edges.

Thanks for your reply. I am not familiar with R. So I further check the codes of bnlearn and pgmpy. I kinda know what you want to do in function '_white_black_list_filter' in structure_learning.py. It seems you want to remove unrelated nodes (such as not in the whitelist). But  '_white_black_list_filter'  goes wrong if whitelist is like [('a', 'b'), ('c', 'd')]. This kind of whiltelist for edges is supported by pgmpy (the base of bnlearn).
So I changed the following part of _white_black_list_filter`

```
if white_list is not None:
            if verbose>=3: print('[bnlearn] >Filter variables on white_list..')
            white_list = [x.lower() for x in white_list]
            Iloc = np.isin(df.columns.str.lower(), white_list)
            df = df.loc[:, Iloc]
```
to

```
if white_list is not None:
            if verbose>=3: print('[bnlearn] >Filter variables on white_list..')
            parent = [ u  for (u, v) in white_list]
            child = [ v  for (u, v) in white_list]
            white_list_node = [x.lower() for x in set(parent+child)]
            Iloc = np.isin(df.columns.str.lower(), white_list_node)
            df = df.loc[:, Iloc]
```
However, changes for blacklist are more complicated. Since only blacklist certain edges do not need to remove nodes. Since I currently, do not need remove nodes. So I will not use function _white_black_list_filter.

Hope this will be helpful for you to make modifications.
By the way, could you please keep the 'fixed_edges' parameter in pgmpy. I think this is helpful. I will try to add it tomorrow.

**Comment by erdogant** [comment]
Im going to look into this!

**Comment by erdogant** [comment]
I made the changes for the filtering of edges and nodes.

First of all, I changed the `bw_list_method ` input for clarity.
`bw_list_method='enforce' `is changed into `bw_list_method='edges'`
`bw_list_method='filter'` is changed into `bw_list_method='nodes'`

However, the filtering on edges only works for `methodtype='hc'` for now.
In addition I included `fixed_edges` but this only works when using `methodtype='hc'`

`pip install -U bnlearn
`
The version should be >= 4.0.0
